### PR TITLE
Improve device enrollment and deletion

### PR DESCRIPTION
**Pull request by DIOHz0r**
Hello,

This pull request is for resolve the issue #32 allowing devices to enroll again once they are deleted and this will solve issue #172 about enrollment process.

Regards.

### CODE REVIEW

File: inc/agent.class.php [hunk 1]
```diff
@@ -588,22 +588,27 @@ public function pre_deleteItem() {
...
       ], true);
 
+      // Send unrolling message to reset agent status
+      if ($this->fields['enroll_status'] == 'enrolled') {
+         $this->sendUnenrollQuery();
+      }
+
       // Delete the MQTT user for the agent
-      if (!empty($serial)) {
+      /*if (!empty($serial)) {
```
Comment (btry) [inline inc/agent.class.php]: I think we shall delete the MQTT account because the agent is deleted.

File: inc/agent.class.php [hunk 2]
```diff
@@ -1091,12 +1096,17 @@ protected function enrollByInvitationToken($input) {
...
-      if (count($rows) > 0) {
-         $event = __('The serial already exists', 'flyvemdm');
-         $this->filterMessages($event);
-         $this->logInvitationEvent($invitation, $event);
-         return false;
+      $computers = $computer->find("`entities_id`='$entityId' AND `serial`='$serial'");
+      if (count($computers) > 0) {
+         // A computer with the same serial exist, let's find if an agent exist for that device.
+         $flyvemdmAgent = new PluginFlyvemdmAgent();
+         $agents = $flyvemdmAgent->find("`entities_id`='$entityId' AND `computers_id`='{$computer->getID()}'");
```
Comment (btry) [inline inc/agent.class.php]: {$computer->getID()} <=== the syntax {$obj->method()} is too messy. I mrefer using a var to store the value of the method then use it here.

File: inc/agent.class.php [hunk 3]
```diff
@@ -1167,20 +1177,32 @@ protected function enrollByInvitationToken($input) {
...
-         'serial'       => $serial,
-         'uuid'         => $uuid,
-      ]);
-      if ($computerId === false) {
-         $event = __("Cannot create the device", 'flyvemdm');
-         $this->filterMessages($event);
-         $this->logInvitationEvent($invitation, $event);
-         return false;
+      // Create/update the device
+      if (count($computers) > 0) {
```
Comment (btry) [inline inc/agent.class.php]: what if count of computers > 1 ?

Which one will you choose ?

**Comment by btry** [comment]
Hi

Globally this workflow may fail. I'm more and more in favor of an import of the computer from the inventory XML. 

I suggest isntead of patching the enrollemnt and deletion workflows we use the import rules engine. This way, the admin of Flyve MDM may choose what happens and use more sophisticated rules to match a enrolling device to an existing computer.

File: inc/agent.class.php (hunk 1 above)
Comment (DIOHz0r) [reply inc/agent.class.php]: ok

File: inc/agent.class.php (hunk 2 above)
Comment (DIOHz0r) [reply inc/agent.class.php]: ok

File: inc/agent.class.php (hunk 1 above)
Comment (btry) [reply inc/agent.class.php]: When we unenroll a device the deletion of the agent occurs when the ACK is received.

see https://github.com/flyve-mdm/flyve-mdm-glpi-plugin/blob/develop/inc/mqtthandler.class.php#L124

**Comment by DIOHz0r** [comment]
Hello,

Travis is passing the test, however there's an [error](https://travis-ci.org/flyve-mdm/glpi-plugin/jobs/319848860#L1024) with the test of Agent integration. 

One of the test on the Agent is making the rest of the tests unchecked and Travis is not taking this as an invalid test.

I've run the test locally and one test is giving and Exception and other one is triggering the empty log problem and stoping the rest of the tests.

I'll be checking and debuging this problem.

**Comment by DIOHz0r** [comment]
Hello @btry,

I tested this change locally with one empty DB and it success so please accept it to merge.

**Pull request merged**